homeassistant/components/axis.py

```diff
@@ -202,69 +209,71 @@ def vapix_service(call):
         """Service to send a message."""
         for _, device in AXIS_DEVICES.items():
             if device.name == call.data[CONF_NAME]:
-                response = device.do_request(call.data[SERVICE_CGI],
-                                             call.data[SERVICE_ACTION],
-                                             call.data[SERVICE_PARAM])
+                response = device.vapix.do_request(call.data[SERVICE_CGI],
+                                                   call.data[SERVICE_ACTION],
+                                                   call.data[SERVICE_PARAM])
                 hass.bus.async_fire(SERVICE_VAPIX_CALL_RESPONSE, response)
                 return True
         _LOGGER.info("Couldn\'t find device %s", call.data[CONF_NAME])
         return False
 
     # Register service with Home Assistant.
-    hass.services.register(DOMAIN,
-                           SERVICE_VAPIX_CALL,
-                           vapix_service,
-                           descriptions[DOMAIN][SERVICE_VAPIX_CALL],
-                           schema=SERVICE_SCHEMA)
-
+    hass.services.async_register(DOMAIN,
+                                 SERVICE_VAPIX_CALL,
+                                 vapix_service,
+                                 descriptions[DOMAIN][SERVICE_VAPIX_CALL],
+                                 schema=SERVICE_SCHEMA)
     return True
 
 
 def setup_device(hass, config, device_config):
     """Set up device."""
     from axis import AxisDevice
 
-    device_config['hass'] = hass
-    device = AxisDevice(device_config)  # Initialize device
-    enable_metadatastream = False
+    def signal_callback(action, event):
```

Comment:
Is this method run on the event loop by axis? If so, please annotate with `@callback`  (imported from `homeassistant.core`)

Reply:
The function is not decorated with coroutine, it is originating from a asyncio datagram endpoint https://github.com/Kane610/axis/blob/7c143aebe2b03c3b7d2c2722dea947f27efe1dfa/axis/rtsp.py#L168
and then finally called from here https://github.com/Kane610/axis/blob/7c143aebe2b03c3b7d2c2722dea947f27efe1dfa/axis/event.py#L85

Reply:
To my understanding, the datagram endpoint is callback based so I guess it is not run on the event loop

Reply:
The datagram endpoint will be called on the event loop, so it's safe to annotate it as `@callback`.